File: ruby/bob/bob.rb
class Bob
  def hey(remark)
    @remark = remark
    if yelled_at?
      'Whoa, chill out!'
    elsif asked_question?
      'Sure.'
    elsif getting_silent_treatment?
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end

  private

  def yelled_at?
    @remark =~ /[A-Z]/ && @remark.upcase == @remark
  end

  def asked_question?
    @remark.end_with?('?')
  end

  def getting_silent_treatment?
    @remark.strip.empty?
  end
end

File: python/bob/bob.py
#
# Skeleton file for the Python "Bob" exercise.
#
import re

def hey(what):
    if silent_treatment(what):
        return 'Fine. Be that way!'
    elif yelled_at(what):
        return 'Whoa, chill out!'
    elif asked_question(what):
        return 'Sure.'
    else:
        return 'Whatever.'

def silent_treatment(what):
    return re.search('\A\s*\Z', what)

def yelled_at(what):
    upper_what = what.upper()
    return re.search('[A-Z]', upper_what) and upper_what == what

def asked_question(what):
    return re.search('\?\Z', what.rstrip())
